proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,200 @@
+assesses the amount of stake that recognizes the feature and decides whether
+it is safe to activate. This is obviously brittle and subject to human error.
+
+If instead the runtime was to assess this stake support for activating a
+feature, this would eliminate the key-holder's responsibility to asses stake
+support, reducing the risk of human error.
+
+In a world where multiple clients will aim to push features and seek to agree on
+their activation, a more automated and secure process will be extremely
+beneficial.
+
+## New Terminology
+
+- **Feature Gate program:** The Core BPF program introduced in
+  [SIMD 0089](https://github.com/solana-foundation/solana-improvement-documents/pull/89)
+  that will own all feature accounts.
+- **Staged Features PDA:** The PDA under the Feature Gate program used to track
+  features submitted for activation for a particular epoch.
+- **Support Signal PDA:** The PDA under the Feature Gate program used to store
+  a bit mask of the staged features a node supports.
+- **Feature Tombstone:** A static address with no on-chain account for assigning
+  accounts under, effectively "archiving" them and removing them from the
+  Feature Gate program's owned accounts.
+
+## Detailed Design
+
+The new process is comprised of the following steps:
+
+1. **Feature Creation:** Contributors create feature accounts as they do now.
+2. **Staging Features for Activation:** In some epoch N, a multi-signature
+   authority stages for activation some or all of the features created in step
+   1.
+3. **Signaling Support for Staged Features:** During the next epoch (epoch N+1),
+   validators signal which of the staged feature-gates they support in their
+   software.
+4. **Activation and Garbage Collection:** At the end of epoch N+1, the runtime
+   activates the feature-gates that have the necessary stake support. At the
+   same time, the runtime also archives activated feature accounts and PDAs no
+   longer required by the process.
+
+### Step 1: Feature Creation
+
+The first step is creation of a feature account, done by submitting a
+transaction containing System instructions to fund, allocate, and assign the
+feature account to `Feature111111111111111111111111111111111111`.
+
+### Step 2: Staging Features for Activation
+
+A multi-signature authority, comprised of key-holders from Solana Labs and
+possibly other validator client teams in the future, will have the authority to
+stage created features for activation.
+In the future, this authority could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction. This instruction
+expects:
+
+- Data: The feature ID
+- Accounts:
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer 
+
+`StageFeatureForActivation` will add the provided feature ID to the **next
+epoch's** Staged Features PDA.
```

Comment:
What if we need to activate a feature urgently, for a security fix?

Reply:
Not sure if this is out of scope or not, but one thought I had was that nodes could regularly check to see if their supported features are different to what their Support Signal PDA indicates, and send a transaction if this is the case.

Reply:
In order to add a new feature gate, the software would need to be updated, so the nodes would restart and then send another transaction with their support signal, so that new feature would be covered.

As far as bypassing the required threshold, I think we'd need a way to set the minimum required threshold per-feature, with the default being 95%. I had this as a side note for a future proposal, but if it's relevant enough I think it can be added to this one.